storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,670 @@
+  // # of attributes being indexed, not tuples.
+  return 1;
+}
+
+bool SMAIndexSubBlock::bulkAddEntries(const TupleIdSequence &tuples) {
+  header_->index_consistent = false;
+  return true;  // There will always be space for the entry.
+}
+
+void SMAIndexSubBlock::bulkRemoveEntries(const TupleIdSequence &tuples) {
+  header_->index_consistent = false;
+}
+
+bool SMAIndexSubBlock::addEntry(const tuple_id tuple) {
+  header_->index_consistent = false;
+  return true;  // There will always be space to insert the entry.
+}
+
+void SMAIndexSubBlock::removeEntry(const tuple_id tuple) {
+  header_->index_consistent = false;
+}
+
+bool SMAIndexSubBlock::rebuild() {
+  resetEntries();
+  header_->count_aggregate = 0;
+  if (tuple_store_.isPacked()) {
+    for (tuple_id tid = 0; tid <= tuple_store_.getMaxTupleID(); ++tid) {
+      addTuple(tid);
+    }
+  } else {
+    for (tuple_id tid = 0; tid <= tuple_store_.getMaxTupleID(); ++tid) {
+      if (tuple_store_.hasTupleWithID(tid)) {
+        addTuple(tid);
+      }
+    }
+  }
+  header_->index_consistent = true;
+  return true;
+}
+
+void SMAIndexSubBlock::addTuple(tuple_id tuple) {
+  for (std::size_t index = 0; index < indexed_attributes_; ++index) {
+    SMAEntry *entry = entries_ + index;
+    TypedValue tuple_value = tuple_store_.getAttributeValueTyped(tuple, entry->attribute);
+
+    // Ignore all nulls.
+    if (tuple_value.isNull()) {
+      continue;
+    }
+
+    if (sma_internal::canSum(entry->type_id)) {
+      entry->sum_aggregate = add_operations_[entry->type_id]->applyToTypedValues(tuple_value, entry->sum_aggregate);
+    }
+
+    // If the entries are valid, we can do comparison, otherwise, we skip the
+    // comparison and set the value.
+    if (!entry->min_entry_ref.valid) {
+      memset(&entry->min_entry_ref.value, 0, sizeof(TypedValue));
+      entry->min_entry_ref.value = tuple_value;
+      entry->min_entry_ref.value.ensureNotReference();
+      entry->min_entry_ref.entry_ref_tuple = tuple;
+      entry->min_entry_ref.valid = true;
+    } else {
+      if (less_comparisons_[entry->type_id]->compareTypedValues(tuple_value, entry->min_entry_ref.value)) {
+        entry->min_entry_ref.value = tuple_value;
+        entry->min_entry_ref.value.ensureNotReference();
+        entry->min_entry_ref.entry_ref_tuple = tuple;
+      }
+    }
+
+    if (!entry->max_entry_ref.valid) {
+      memset(&entry->max_entry_ref.value, 0, sizeof(TypedValue));
+      entry->max_entry_ref.value = tuple_value;
+      entry->max_entry_ref.value.ensureNotReference();
+      entry->max_entry_ref.entry_ref_tuple = tuple;
+      entry->max_entry_ref.valid = true;
+    } else {
+      if (less_comparisons_[entry->type_id]->compareTypedValues(entry->max_entry_ref.value, tuple_value)) {
+        entry->max_entry_ref.value = tuple_value;
+        entry->max_entry_ref.value.ensureNotReference();
+        entry->max_entry_ref.entry_ref_tuple = tuple;
+      }
+    }
+  }
+
+  header_->count_aggregate++;
```

Comment:
Yup, so it might make sense to add a count field to an SMAEntry because, with nulls, individual attribute counts will be different from the relation's COUNT. I'll add this as a TODO